modules/katamari/src/main/ts/ephox/katamari/api/Maybe.ts

```diff
@@ -0,0 +1,207 @@
+export type Maybe<T> = Just<T> | Nothing;
+
+export interface Just<T> {
+  readonly tag: 'JUST';
+  readonly value: T;
+}
+
+export interface Nothing {
+  readonly tag: 'NOTHING';
+}
+
+// Constructors
+export const just = <T>(value: T): Just<T> => ({
```

Comment:
Worth mentioning: at present this can cause some issues with tsc inference. The problem revolves around "widening", explained below.

Typescript will automatically **widen** `"specific-string"` to `string` in it's type checking **in certain situations**. These situations include things like:
- when you use a `let` binding instead of const (`let thingo: "specific-string" = "specific-string"` makes it a bit hard to mutate)
- when you put in in an array with other strings
- when you have an expression that will conditionally evaluate to either `"specific-string"` or a different string
(note that in all of these situations you can explicitly ask typescript to not widen the type, if you choose)

Because tsc can automate the widening process at the times you'd want to widen things, they can safely default to the narrowest types possible, and trust that instead of something like this:
```typescript
type "other-string" not assignable to type "specific-string"
```
you won't get any errors (as the type widener has inferred that it's just a `string`)

Typescript will do similar widening for array literals (`[string, number] => (string | number)[]`) and for numeric literals (`3 => number`) where it sees fit. Unfortunately, we can't persuade typescript to do automatic type widening for custom types.

Consider the following code:
```typescript
let state = Maybe.just(3);
// do things
state = Maybe.bind(state, (inner) => ...)
```

This code won't compile, because `state` is of type `Just<number>` instead of type `Maybe<number>`, and therefore trying to assign a `Maybe<number>` to it (the return type of bind) will fail under the claim `type 'Nothing' is not assignable to type 'Just<number>'`. In an ideal world, Typescript would see that this was a `let` binding, and automatically assign the type to `Maybe<number>` instead of `Just<number>`. That's not an option, which leaves us with the following two choices:

## Choice 1: Manual type widening
By being explicit about the types, we can tell Typescript that things are widened manually, sidestepping the "how do we infer it" problem. This would involve the following types of changes:
- `let state = Maybe.just(3)` ==> `let state: Maybe<number> = Maybe.just(3)`
- Consider this hypothetical code:
```typescript
const toMaybe = <T>(opt: Optional<T>): Maybe<T> => opt.fold(Fun.constant(Maybe.Nothing), Maybe.just)
```
We would need to be change `opt.fold(...` to `opt.fold<Maybe<T>>(...` lest the compiler complain that the two arguments of `fold` return different types.
- `const x = [just(3), just(5), nothing]` ==> `const x: Array<Maybe<number>> = [just(3), just(5), nothing]`

## Choice 2: Initial type widening
Rather than having the `just` function returning the `Just` type, we could just make it return `Maybe`. This would avoid all of those problems, but it would also make things less specific / less technically correct. Hopefully, though, the type of inferences this change would break ("hey this `Maybe` was created with the `just` constructor, so it'll always have a value") don't come up in real code (if there's no conditionality in the code, or you're guaranteed to be using the same constructor any time that code is executed, why are you using "Maybe" at all?).

## Choice 3: See if upstream Typescript supports this, or make a request for them to support this
I don't think this is a great idea, but I thought I'd mention it in case someone else thinks we should do this. Choice 2 seems like a good solution for us (although it is technically suboptimal, it is practically usable), so I don't see the need for implementing this _for our usecase_.